### PR TITLE
refactor(contracts): cleanup unused scripts and organize ts code

### PR DESCRIPTION
**Pull request by ctrlc03**
- [x] Remove unused scripts in the contracts folder
- [x] refactor and add typedoc comments to the typescript code 
- [x] ensure we can pass custom signers to the deployment functions.

Related to #843

### CODE REVIEW

**Comment by kittybest** [comment]
https://github.com/privacy-scaling-explorations/maci/blob/efe8d5f82e0c0705a8d311488c47c64a08fa05dd/contracts/tests/utils.ts#L129
Should the `contract` here be changed to `aqContract` to keep consistence with other testFill functions?

**Comment by kittybest** [comment]
https://github.com/privacy-scaling-explorations/maci/blob/efe8d5f82e0c0705a8d311488c47c64a08fa05dd/contracts/tests/AccQueue.test.ts#L38

This has been tested in `testEmptyUponDeployment` above, not sure should it be tested again?
If we just wanna test the contract couldn't get an non-existing index, maybe choose another index instead of 0?

**Comment by kittybest** [comment]
https://github.com/privacy-scaling-explorations/maci/blob/efe8d5f82e0c0705a8d311488c47c64a08fa05dd/contracts/tests/utils.ts#L188

For me, I would expect this function testing if enqueueing leaves works well. Not sure why we need to insert another subtree? Isn't it testing two things? (The leaves could be successfully inserted, and if the tree is full, there will be another subtree) or actually they are the same thing? (please tell me if I understand wrong)

File: contracts/tests/AccQueueBenchmark.test.ts [hunk 1]
```diff
@@ -1,6 +1,7 @@
 import { expect } from "chai";
 import { AccQueue, NOTHING_UP_MY_SLEEVE } from "maci-crypto";
 
+import { getDefaultSigner } from "../ts";
 import { deployPoseidonContracts, linkPoseidonLibraries } from "../ts/deploy";
```
[the remark this answers is not in the document]
Comment (ctrlc03) [reply contracts/tests/AccQueueBenchmark.test.ts:4]: I changed every test to import functions from the actual file as it looks like it's what we are doing mostly across the repo (being consistent would be great) - good catch on this

**Comment by ctrlc03** [comment]
> https://github.com/privacy-scaling-explorations/maci/blob/efe8d5f82e0c0705a8d311488c47c64a08fa05dd/contracts/tests/utils.ts#L129
> 
> 
> Should the `contract` here be changed to `aqContract` to keep consistence with other testFill functions?

Sure, changed that for all accQueue test functions

**Comment by kittybest** [comment]
https://github.com/privacy-scaling-explorations/maci/blob/282ff2809e6aa050eefacd84233ba8268a299cf1/contracts/tests/AccQueue.test.ts#L467

The test here is actually being tested also in the edge tests
https://github.com/privacy-scaling-explorations/maci/blob/282ff2809e6aa050eefacd84233ba8268a299cf1/contracts/tests/AccQueue.test.ts#L207

**Comment by kittybest** [comment]
https://github.com/privacy-scaling-explorations/maci/blob/724f18ed880d051b8f1d5beb1689770e625c9b22/contracts/tests/Tally.test.ts#L129
Just curious that should we make arguments in `packTallyVotesSmallVals` function same as the `tallyContract.genTallyVotesPackedVals`?

**Comment by kittybest** [comment]
https://github.com/privacy-scaling-explorations/maci/blob/724f18ed880d051b8f1d5beb1689770e625c9b22/contracts/tests/MessageProcessor.test.ts#L148C13-L148C13
why we use `toString(16)` here while using just `toString()` elsewhere?

**Comment by kittybest** [comment]
https://github.com/privacy-scaling-explorations/maci/blob/724f18ed880d051b8f1d5beb1689770e625c9b22/contracts/tests/MaciOverflow.test.ts#L28
what kind of overflow are we testing here?

**Comment by kittybest** [comment]
https://github.com/privacy-scaling-explorations/maci/blob/724f18ed880d051b8f1d5beb1689770e625c9b22/contracts/tests/MaciOverflow.test.ts#L54-L55
should we add how the numbers calculated in the comments?

**Comment by kittybest** [comment]
https://github.com/privacy-scaling-explorations/maci/blob/724f18ed880d051b8f1d5beb1689770e625c9b22/contracts/package.json#L16-L25
I think we miss the `test-hasherBenchmarks` and `test-maciOverflow` here?

**Comment by kittybest** [comment]
https://github.com/privacy-scaling-explorations/maci/blob/724f18ed880d051b8f1d5beb1689770e625c9b22/contracts/tests/MaciOverflow.test.ts#L30
https://github.com/privacy-scaling-explorations/maci/blob/724f18ed880d051b8f1d5beb1689770e625c9b22/contracts/tests/SignUpGatekeeper.test.ts#L73
I tried to remove the `as { x: BigNumberish; y: BigNumberish }` part to make it clearer, and it still works, so I think we could remove this?

**Comment by kittybest** [comment]
https://github.com/privacy-scaling-explorations/maci/blob/724f18ed880d051b8f1d5beb1689770e625c9b22/contracts/tests/MACI.test.ts#L230-L233
Do you know why put the first `BigInt(0)` in the array, while adding other BigInt(0) in the for loop?
Same as the below parts:
https://github.com/privacy-scaling-explorations/maci/blob/724f18ed880d051b8f1d5beb1689770e625c9b22/contracts/tests/MessageProcessor.test.ts#L85-L88
https://github.com/privacy-scaling-explorations/maci/blob/724f18ed880d051b8f1d5beb1689770e625c9b22/contracts/tests/Poll.test.ts#L72-L75
https://github.com/privacy-scaling-explorations/maci/blob/724f18ed880d051b8f1d5beb1689770e625c9b22/contracts/tests/Tally.test.ts#L84-L87

**Comment by kittybest** [comment]
Do we test to prevent accidentally input negative numbers somewhere? 
And I found most of the gasLimit are different, how are they decided?

**Comment by kittybest** [comment]
https://github.com/privacy-scaling-explorations/maci/blob/724f18ed880d051b8f1d5beb1689770e625c9b22/contracts/tests/MACI.test.ts#L246
another question, why we need the `init()` function public for poll contract? it seems to be inited once the maci contract call `deployPoll()`.

**Comment by ctrlc03** [comment]
> Do we test to prevent accidentally input negative numbers somewhere? And I found most of the gasLimit are different, how are they decided?

All numeric variables we use in the smart contracts are `uint` so no negative values can be accepted.
Honestly I don't know how they decided about the gas limits for the tests, probably just put what worked at the time to get the test through (some being more expensive computations than others and requiring more gas to complete). I'll work today on refactoring the actual tests and address some of the points you brought up about the tests

**Comment by ctrlc03** [comment]
> https://github.com/privacy-scaling-explorations/maci/blob/724f18ed880d051b8f1d5beb1689770e625c9b22/contracts/tests/MACI.test.ts#L246
> 
> 
> another question, why we need the `init()` function public for poll contract? it seems to be inited once the maci contract call `deployPoll()`.

So you suggest we make it `external` instead? that would definitely work

**Comment by ctrlc03** [comment]
> https://github.com/privacy-scaling-explorations/maci/blob/724f18ed880d051b8f1d5beb1689770e625c9b22/contracts/tests/MACI.test.ts#L230-L233
> 
> 
> Do you know why put the first `BigInt(0)` in the array, while adding other BigInt(0) in the for loop?
> Same as the below parts:
> https://github.com/privacy-scaling-explorations/maci/blob/724f18ed880d051b8f1d5beb1689770e625c9b22/contracts/tests/MessageProcessor.test.ts#L85-L88
> 
> 
> https://github.com/privacy-scaling-explorations/maci/blob/724f18ed880d051b8f1d5beb1689770e625c9b22/contracts/tests/Poll.test.ts#L72-L75
> 
> 
> https://github.com/privacy-scaling-explorations/maci/blob/724f18ed880d051b8f1d5beb1689770e625c9b22/contracts/tests/Tally.test.ts#L84-L87

Not sure, seems unnecessary, I'll address this when I work on the test suite

**Comment by ctrlc03** [comment]
> https://github.com/privacy-scaling-explorations/maci/blob/724f18ed880d051b8f1d5beb1689770e625c9b22/contracts/tests/MaciOverflow.test.ts#L54-L55
> 
> 
> should we add how the numbers calculated in the comments?

definitely!

**Comment by ctrlc03** [comment]
> https://github.com/privacy-scaling-explorations/maci/blob/724f18ed880d051b8f1d5beb1689770e625c9b22/contracts/tests/MaciOverflow.test.ts#L28
> 
> 
> what kind of overflow are we testing here?

mm agree this test case is a bit vague, will refactor it when working on the test suite

**Comment by ctrlc03** [comment]
> n

Good shout, definitely we should be as consistent as possible

**Comment by kittybest** [comment]
> > https://github.com/privacy-scaling-explorations/maci/blob/724f18ed880d051b8f1d5beb1689770e625c9b22/contracts/tests/MACI.test.ts#L246
> > 
> > another question, why we need the `init()` function public for poll contract? it seems to be inited once the maci contract call `deployPoll()`.
> 
> So you suggest we make it `external` instead? that would definitely work

I'm not sure whether make it as an `external` is better or not. 
Just wonder why we need the `init` function for `Poll.sol` since it looks like just setting up `isInit` to `true`? and if it's called more than once, the call would be reverted, but currently I think the only one who will call this is the `PollFactory`? so why not just put the command of set `isInit` to `true` in the constructor, removing the possibility of user accidentally calling `poll.init()`?

File: contracts/tests/utils.ts [hunk 2]
```diff
@@ -155,24 +169,24 @@ export const testFillForAllIncompletes = async (
...
   expect(numLeaves.toString()).to.equal("0");
 
-  await expect(contract.getSubRoot(0)).to.be.revertedWithCustomError(contract, "InvalidIndex");
+  await expect(aqContract.getSubRoot(0)).to.be.revertedWithCustomError(aqContract, "InvalidIndex");
 };
 
 /**
  * Enqueue leaves and check their subroots
- * @param contract - the AccQueue contract
+ * @param aqContract - the AccQueue contract
```
Comment (samajammin) [inline contracts/tests/utils.ts:183]: 👍

**Comment by samajammin** [comment]
> * [x]  ensure we can pass custom signers to the deployment functions.

Curious, why do we want this?

**Comment by kittybest** [comment]
https://github.com/privacy-scaling-explorations/maci/blob/458bc6b09cbc8db652dbb72525e299a4c2631caf/contracts/ts/deploy.ts#L350-L360
Just curious about why we assign the Poseidon contracts in the deploy function arguments for other contracts deploying, such as `deploySubsidy`, `deployTally`, etc, while we deploy Poseidon and link the libraries independently in the `deployMaci` function?

**Comment by ctrlc03** [comment]
> > * [x]   ensure we can pass custom signers to the deployment functions.
> 
> Curious, why do we want this?

This opens up more testing scenarios, where you can deploy smart contracts from different addresses rather than all from the same. Also you could use the signer of this custom deployer class https://github.com/privacy-scaling-explorations/maci/blob/458bc6b09cbc8db652dbb72525e299a4c2631caf/contracts/ts/deployer.ts#L20 which could fit other use cases

**Comment by ctrlc03** [comment]
> > * [x]   ensure we can pass custom signers to the deployment functions.
> 
> Curious, why do we want this?



> > > https://github.com/privacy-scaling-explorations/maci/blob/724f18ed880d051b8f1d5beb1689770e625c9b22/contracts/tests/MACI.test.ts#L246
> > > 
> > > another question, why we need the `init()` function public for poll contract? it seems to be inited once the maci contract call `deployPoll()`.
> > 
> > 
> > So you suggest we make it `external` instead? that would definitely work
> 
> I'm not sure whether make it as an `external` is better or not. Just wonder why we need the `init` function for `Poll.sol` since it looks like just setting up `isInit` to `true`? and if it's called more than once, the call would be reverted, but currently I think the only one who will call this is the `PollFactory`? so why not just put the command of set `isInit` to `true` in the constructor, removing the possibility of user accidentally calling `poll.init()`?

So the problem with that is that we need an `init` function inside Poll because there is a initialization step we must do before it is usable, but cannot be done inside the constructor (at this time).
Basically we first deploy a Poll contract inside the PollFactory contract. Then we deploy an AccQueue instance, and then we transfer ownership of the AccQueue instance to the Poll contract. For that, we need the address of the Poll contract which we only have after deployment (of course there are other patterns in Solidity which can be used to deploy smart contracts and for instance with [create2](https://docs.openzeppelin.com/cli/2.8/deploying-with-create2) it's possible to know the address before the deployment, though this is not something we have implemented atm and within the scope of this refactoring). As you can see, inside the `init` function of the Poll we `enqueue` a message into the `messageAq` instance, which we can only do when the Poll is the owner of `messageAq`.

**Review by kittybest — APPROVED** [review]

**Pull request merged**